echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/TriggerRepository.groovy

```diff
@@ -0,0 +1,62 @@
+package com.netflix.spinnaker.echo.scheduler.actions.pipeline.impl
+
+import com.netflix.spinnaker.echo.model.Pipeline
+import com.netflix.spinnaker.echo.model.Trigger
+import groovy.util.logging.Slf4j
+
+import javax.annotation.Nullable
+
+@Slf4j
+class TriggerRepository {
+  private final Map<String, Trigger> triggersById
+
+  TriggerRepository(List<Pipeline> pipelines) {
+    triggersById = new HashMap<>()
+    for (Pipeline pipeline : pipelines) {
+      for (Trigger trigger : pipeline.triggers) {
+        if (trigger.id == null) {
+          // see PipelineCache::decorateTriggers
+          throw new IllegalStateException("Trigger with null id ${trigger}")
+        }
+
+        Trigger previous = triggersById.put(trigger.id, trigger)
+        if (previous) {
+          log.warn("Duplicate trigger ids found: ${previous} with parent ${previous.parent} and ${trigger} with parent ${trigger.parent}")
```

Comment:
similarly here ... you're regenerating trigger ids so how could you end up with a previous one?

(having a check like this implies that there could be some reason for a duplicate trigger which may prove confusing months/years down the line)

Reply:
This is just defensive coding: it's not supposed to happen, but it doesn't cost much to check and I'd rather have a log than fail silently

Reply:
throw an exception ... emit a metric ... log is just going to get lost in the noise if assumptions ever break.